pkg/execution/executor/executor.go

```diff
@@ -450,6 +498,205 @@ func (e *executor) executeStep(ctx context.Context, id state.Identifier, step *i
 	return response, idx, err
 }
 
+// HandlePauses handles pauses loaded from an incoming event.
+//
+// TODO: Use InternalEvent which wraps event with internal ID here.
+func (e *executor) HandlePauses(ctx context.Context, iter state.PauseIterator, evt event.TrackedEvent) error {
+	if e.queue == nil || e.sm == nil {
+		return fmt.Errorf("No queue or state manager specified")
+	}
+
+	var (
+		goerr error
+		wg    sync.WaitGroup
+	)
+
+	evtID := evt.InternalID()
+
+	// Schedule up to PauseHandleConcurrency pauses at once.
+	sem := semaphore.NewWeighted(int64(PauseHandleConcurrency))
+
+	for iter.Next(ctx) {
+		if goerr != nil {
+			break
+		}
+
+		pause := iter.Val(ctx)
+
+		// Block until we have capacity
+		_ = sem.Acquire(ctx, 1)
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			// Always release one from the capacity
+			defer sem.Release(1)
+
+			if pause.TriggeringEventID != nil && *pause.TriggeringEventID == evt.InternalID().String() {
+				// Don't allow the original function trigger to trigger the
+				// cancellation
+				return
+			}
+
+			// NOTE: Some pauses may be nil or expired, as the iterator may take
+			// time to process.  We handle that here and assume that the event
+			// did not occur in time.
+			if pause == nil || pause.Expires.Time().Before(time.Now()) {
+				// Consume this pause to remove it entirely
+				_ = e.sm.ConsumePause(context.Background(), pause.ID, nil)
```

Comment:
`ConsumePause` seems like it always need to called before a return.
Could it be wrapped in a `defer` to reduce explicit calls everywhere?

Reply:
Almost always, though sometimes we can error eg. if creating a new expression evaluator, or if there's somehow an error in the state store cancelling the function.  In these cases we don't want to consume the pause - we'd hope that the error is transient and will retry.